bugfix #39505: TypeError in SelfRegistration with reg-code & local roles

utesche opened the pull request:
The PR fixes 3 TypeErrors for
issue:  https://mantis.ilias.de/view.php?id=39505
forum: https://docu.ilias.de/goto_docu_frm_1875_8811.html
.

utesche committed:
bugfix #39505: TypeError in SelfRegistration with reg-code & local role(s)

The PR fixes 3 TypeErrors for
issue:  https://mantis.ilias.de/view.php?id=39505
forum: https://docu.ilias.de/goto_docu_frm_1875_8811.html
.

mjansenDatabay commented:
Thx for the PR @utesche . I suggest to cast the values in the variable assingment.

Instead of:

```php
        $code_local_roles = [];
```

AND

```php
                $code_data = ilRegistrationCode::getCodeData($code);
                if ($code_data["role_local"]) {
                    // need user id before we can assign role(s)
                    $code_local_roles = explode(";", $code_data["role_local"]);
                }
```

Should: All values of the `$code_local_roles` list should be of type `int`.

```php
        /** @var list<int>|null $code_local_roles */
        $code_local_roles = [];
```
AND

```php
                $code_data = ilRegistrationCode::getCodeData($code);
                if (isset($code_data['role_local']) && is_string($code_data['role_local'])) {
                    // need user id before we can assign role(s)
                    $code_local_roles = array_filter(array_map(
                        static fn (string $value): int => (int) $value,
                        explode(';', $code_data['role_local'])
                    ));
                }
```

Whenever `$code_local_roles` is used, the consuming code can expect all values to be of type `int`.

---

Of course, it is still not nice that consumers (like `class.ilAccountRegistrationGUI.php`) have to know internals (separator) of `$code_data['role_local']`. But that should not be part of this PR, this requires some deep refactorings.

Best regards,
Michael

mjansenDatabay commented:
@utesche Do have time/energy and want to include the review remarks into your PR?

utesche committed:
Update class.ilAccountRegistrationGUI.php

Changed to suggestion by mjansen here:
https://github.com/ILIAS-eLearning/ILIAS/pull/6981#issuecomment-1916379489

utesche commented:
@mjansenDatabay,
as suggested, 
the strict casting to int of separate local_role_ids at usage time is changed now 
to a conversion of the whole array of ids to int when creating the array (by exploding the given input).

mjansenDatabay commented on Services/Registration/classes/class.ilAccountRegistrationGUI.php:
```diff
@@ -575,6 +580,7 @@ protected function createUser(int $a_role): string
         // local roles from code
         if ($this->code_was_used && is_array($code_local_roles)) {
             foreach (array_unique($code_local_roles) as $local_role_obj_id) {
+                $local_role_obj_id = $local_role_obj_id;
```
IMO this line can now be deleted.

```suggestion
```

mjansenDatabay commented:
@pascalseeland Please check my suggested change and merge the PR afterwards, if you don't have any further objections.

pascalseeland committed:
Remove unnecessary assignment

Co-authored-by: Michael Jansen <mjansen@databay.de>
